// expo-cli/build/commands/build/ios/appleApi/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "authenticate", {
  enumerable: true,
  get: function () {
    return _authenticate().default;
  }
});
Object.defineProperty(exports, "ensureAppExists", {
  enumerable: true,
  get: function () {
    return _ensureAppExists().default;
  }
});
Object.defineProperty(exports, "setup", {
  enumerable: true,
  get: function () {
    return _setup().default;
  }
});
exports.createManagers = void 0;

function _authenticate() {
  const data = _interopRequireDefault(require("./authenticate"));

  _authenticate = function () {
    return data;
  };

  return data;
}

function _distributionCert() {
  const data = _interopRequireDefault(require("./distributionCert"));

  _distributionCert = function () {
    return data;
  };

  return data;
}

function _pushKey() {
  const data = _interopRequireDefault(require("./pushKey"));

  _pushKey = function () {
    return data;
  };

  return data;
}

function _provisioningProfile() {
  const data = _interopRequireDefault(require("./provisioningProfile"));

  _provisioningProfile = function () {
    return data;
  };

  return data;
}

function _ensureAppExists() {
  const data = _interopRequireDefault(require("./ensureAppExists"));

  _ensureAppExists = function () {
    return data;
  };

  return data;
}

function _setup() {
  const data = _interopRequireDefault(require("./setup"));

  _setup = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const createManagers = ctx => ({
  distributionCert: (0, _distributionCert().default)(ctx),
  pushKey: (0, _pushKey().default)(ctx),
  provisioningProfile: (0, _provisioningProfile().default)(ctx)
});

exports.createManagers = createManagers;
