// expo-cli/build/PackageManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createForProject = createForProject;
exports.YarnPackageManager = exports.NpmPackageManager = void 0;

function _ansiRegex() {
  const data = _interopRequireDefault(require("ansi-regex"));

  _ansiRegex = function () {
    return data;
  };

  return data;
}

function _config() {
  const data = require("@expo/config");

  _config = function () {
    return data;
  };

  return data;
}

function _spawnAsync() {
  const data = _interopRequireDefault(require("@expo/spawn-async"));

  _spawnAsync = function () {
    return data;
  };

  return data;
}

function _split() {
  const data = _interopRequireDefault(require("split"));

  _split = function () {
    return data;
  };

  return data;
}

function _stream() {
  const data = require("stream");

  _stream = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("./log"));

  _log = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const ansi = `(?:${(0, _ansiRegex().default)().source})*`;
const npmPeerDependencyWarningPattern = new RegExp(`${ansi}npm${ansi} ${ansi}WARN${ansi}.+You must install peer dependencies yourself\\.\n`, 'g');
const yarnPeerDependencyWarningPattern = new RegExp(`${ansi}warning${ansi} "[^"]+" has (?:unmet|incorrect) peer dependency "[^"]+"\\.\n`, 'g');

class NpmStderrTransform extends _stream().Transform {
  _transform(chunk, encoding, callback) {
    this.push(chunk.toString().replace(npmPeerDependencyWarningPattern, ''));
    callback();
  }

}

class YarnStderrTransform extends _stream().Transform {
  _transform(chunk, encoding, callback) {
    this.push(chunk.toString().replace(yarnPeerDependencyWarningPattern, ''));
    callback();
  }

}

class NpmPackageManager {
  constructor({
    cwd
  }) {
    this.options = {
      cwd,
      stdio: ['inherit', 'inherit', 'pipe']
    };
  }

  get name() {
    return 'npm';
  }

  async installAsync() {
    await this._runAsync(['install']);
  }

  async addAsync(...names) {
    await this._runAsync(['install', '--save', ...names]);
  }

  async addDevAsync(...names) {
    await this._runAsync(['install', '--save-dev', ...names]);
  } // Private


  async _runAsync(args) {
    (0, _log().default)(`> npm ${args.join(' ')}`);
    const promise = (0, _spawnAsync().default)('npm', [...args], this.options);

    if (promise.child.stderr) {
      promise.child.stderr.pipe((0, _split().default)(/\r?\n/, line => line + '\n')).pipe(new NpmStderrTransform()).pipe(process.stderr);
    }

    await promise;
  }

}

exports.NpmPackageManager = NpmPackageManager;

class YarnPackageManager {
  constructor({
    cwd
  }) {
    this.options = {
      cwd,
      stdio: ['inherit', 'inherit', 'pipe']
    };
  }

  get name() {
    return 'Yarn';
  }

  async installAsync() {
    await this._runAsync(['install']);
  }

  async addAsync(...names) {
    await this._runAsync(['add', ...names]);
  }

  async addDevAsync(...names) {
    await this._runAsync(['add', '--dev', ...names]);
  } // Private


  async _runAsync(args) {
    (0, _log().default)(`> yarn ${args.join(' ')}`);
    const promise = (0, _spawnAsync().default)('yarnpkg', args, this.options);

    if (promise.child.stderr) {
      promise.child.stderr.pipe(new YarnStderrTransform()).pipe(process.stderr);
    }

    await promise;
  }

}

exports.YarnPackageManager = YarnPackageManager;

function createForProject(projectRoot, options = {}) {
  let PackageManager;

  if (options.npm) {
    PackageManager = NpmPackageManager;
  } else if (options.yarn) {
    PackageManager = YarnPackageManager;
  } else if ((0, _config().isUsingYarn)(projectRoot)) {
    PackageManager = YarnPackageManager;
  } else {
    PackageManager = NpmPackageManager;
  }

  return new PackageManager({
    cwd: projectRoot
  });
}
